src/TTextEdit.cpp

```diff
@@ -2581,3 +2620,128 @@ void TTextEdit::reportCodepointErrors()
         qDebug().nospace().noquote() << " ";
     }
 }
+
+void TTextEdit::setCaretPosition(int line, int column)
+{
+    mCaretLine = line;
+    mCaretColumn = column;
+
+    if (!mudlet::self()->isCaretModeEnabled()) {
+        return;
+    }
+
+    updateCaret();
+}
+
+void TTextEdit::updateCaret()
+{
+    int lineOffset = imageTopLine();
+    if (!mIsLowerPane && mCaretLine < lineOffset) {
+        scrollTo(mCaretLine);
+    } else if (!mIsLowerPane && mCaretLine >= lineOffset + mScreenHeight) {
+        int emptyLastLine = mpBuffer->lineBuffer.last().length() == 0;
+        if (mCaretLine == mpBuffer->lineBuffer.length() - 1 - emptyLastLine) {
+            scrollTo(mCaretLine + 2);
+        } else {
+            scrollTo(mCaretLine + 1);
+        }
+    }
```

Comment:
Refactoring out the `!mIsLowerPane` common term:
```suggestion
    if (!mIsLowerPane) {
        if (mCaretLine < lineOffset) {
            scrollTo(mCaretLine);
        } else if (mCaretLine >= lineOffset + mScreenHeight) {
            int emptyLastLine = mpBuffer->lineBuffer.last().length() == 0;
            if (mCaretLine == mpBuffer->lineBuffer.length() - 1 - emptyLastLine) {
                scrollTo(mCaretLine + 2);
            } else {
                scrollTo(mCaretLine + 1);
            }
        }
    }
```

Reply:
Great! This indeed looks better.

Again, I made this change directly in the commit that added this code. Please let me know if you would prefer me to use the “Commit suggestion” feature. It’s just that I’m used to the workflow of improving the individual commits in a PR branch until they are “perfect”.

Just keep adding commits on top of the branch feels to me like working on a Jenga tower. :)

Reply:
Yeah, but they will all go away when it gets **Squashed-and-merged** (our preferred choice) when the PR is accepted and merged into the `development` branch - hopefully the core-team member will take the time to sensibly combine all the commit messages into a single sensible one at the time.